core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -3154,6 +3155,89 @@ private[spark] class RedirectThread(
   }
 }
 
+/**
+ * PythonRunner process wrapper. In this wrapper will collect process error message,
+ * if process exit with exception, spark can get process exit message.
+ *
+ * @param process PythonRunner python process
+ * @param out Where to redirect the message.
+ * @param name process name.
+ * @param propagateEof If propagate Eof.
+ */
+private[spark] class PythonRunnerProcessWrapper(
```

Comment:
Do we have a reference implementation? If not we need to review this class carefully.